credential/concept/template.go

```diff
@@ -0,0 +1,319 @@
+const IDField = "id"
+
+// IssuerField defines the concept/VC JSON path to a VC issuer
+const IssuerField = "issuer"
+
+// SubjectField defines the concept JSON path to a VC subject
+const SubjectField = "subject"
+
+// Template represents a json mapping template. In the template concept to json path mappings are stored.
+// Indices that need to be created by a DB are also created by the template.
+type Template interface {
+	// Indices returns a list of compound indices. Each entry contains the concept json path.
+	Indices() [][]string
+	// VCType returns the VC type.
+	VCType() string
+	// ToVCPath returns the mapping of concept path to VC path
+	ToVCPath(conceptPath string) string
+}
+
+type template struct {
+	raw string
+	// conceptIndexMapping contains mappings from concept name to VC JSON path
+	conceptIndexMapping map[string]string
+	// reverseIndexMapping contains mappings from VC JSON path to concept JSON path
+	reverseIndexMapping map[string]string
+	// indices contains a set of compound indices
+	indices [][]string
+	// key: value pairs that always need to be added to the backend query
+	fixedValues map[string]string
+}
+
+// templateString represents a value in a template.
+// it contains several convenience methods for parsing purposes
+type templateString string
+
+// ToVCPath returns the mapping of concept path to VC path
+func (ct *template) ToVCPath(conceptPath string) string {
+	return ct.conceptIndexMapping[conceptPath]
+}
+
+func (ct *template) Indices() [][]string {
+	return ct.indices
+}
+
+// VCType returns the VC type.
+func (ct *template) VCType() string {
+	return ct.fixedValues[TypeField]
+}
+
+// find paths to concepts
+func (ct *template) parse() error {
+	jq := gojsonq.New().FromString(ct.raw)
+
+	// type is a special pony
+	ct.conceptIndexMapping = map[string]string{TypeField: TypeField}
+	ct.reverseIndexMapping = map[string]string{TypeField: TypeField}
+	ct.fixedValues = map[string]string{}
+
+	if err := ct.mapR(jq, ""); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ct *template) concepts() []string {
+	var cs = make([]string, 0)
+
+	// standard concepts not needed
+	for k := range ct.conceptIndexMapping {
+		s := strings.Split(k, ".")
+		if len(s) > 1 {
+			cs = append(cs, k)
+		}
+	}
+
+	return cs
+}
+func (ct *template) rootConcepts() []string {
+	var cm = make(map[string]bool, 0)
+
+	// all concepts
+	for _, c := range ct.concepts() {
+		cm[strings.Split(c, ".")[0]] = true
+	}
+
+	// uniq concepts
+	var cs = make([]string, 0)
+	for k := range cm {
+		cs = append(cs, k)
+	}
+
+	return cs
+}
+
+func (ct *template) mapR(jq *gojsonq.JSONQ, currentPath string) error {
+	val := jq.Get()
+
+	if jq.Error() != nil {
+		return jq.Error()
+	}
+
+	// slice
+	if _, ok := val.([]interface{}); ok {
+		return errors.New("json arrays are not supported")
+	}
+
+	if m, ok := val.(map[string]interface{}); ok {
+		for k, v := range m {
+			nextPath := path(currentPath, k)
+
+			if sv, ok := v.(string); ok {
```

Comment:
if this function is recursive, shouldn't this be at top level of the func, next to `slice` and `map`? Maybe it makes testing and reading easier (and lower complexity due to less nested control flows)

Reply:
This was a result of the json parsing. In the end I was able to remove the json library altogether since parsing to a `map[string]interface{}` works equally well. This has the side effect of placing the string check at the top level.